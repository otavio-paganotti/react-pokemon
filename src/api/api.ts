import axios, { AxiosInstance } from 'axios';

export const http: AxiosInstance = axios.create({
  baseURL: String(import.meta.env.VITE_APP_API_URL || ''),
});
