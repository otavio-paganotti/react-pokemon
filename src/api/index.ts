/**
 * Exporting the Axios Instance
 */
export * from './api';

/**
 * Exporting all modules
 */
export * as Pokemon from './services/rest/pokemon';
