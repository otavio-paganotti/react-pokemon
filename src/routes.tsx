import type { RouteObject } from 'react-router';
import MainLayout from '@/layouts/Default';
import Home from '@/pages/index/Index'
import Pokemon from '@/pages/pokemon/Index'

const routes: RouteObject[] = [
  {
    path: '*',
    element: <MainLayout />,
    children: [
      {
        index: true,
        element: <Home />,
      },
      {
        index: false,
        path: ':id',
        element: <Pokemon />
      }
    ],
  },
];

export default routes;