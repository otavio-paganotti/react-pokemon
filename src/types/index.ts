export type Nullable<T = unknown> = T | null;

export interface PokemonItem {
  name: string;
  url: string;
  id: number;
  content: Pokemon;
}

export interface Response<T> {
  count: number;
  next: Nullable<string>;
  previous: Nullable<string>;
  results: T[];
  pagination: {
    firstPage: number;
    lastPage: number;
    nextPage: number;
    prevPage: number;
    nextPages: number[];
    prevPages: number[];
    page: number;
  };
}

export interface Sprites {
  back_default: string;
}

export interface Stat {
  base_stat: number;
  effort: number;
  stat: Species;
}

export interface Type {
  slot: number;
  type: Species;
}

export interface VersionGroupDetail {
  level_learned_at: number;
  move_learn_method: Species;
  version_group: Species;
}

export interface Move {
  move: Species;
  version_group_details: VersionGroupDetail[];
}

export interface VersionDetail {
  rarity: number;
  version: Species;
}

export interface Species {
  name: string;
  url: string;
}

export interface HeldItem {
  item: Species;
  version_details: VersionDetail[];
}

export interface GameIndex {
  game_index: number;
}

export interface Ability {
  ability: Species;
  is_hidden: boolean;
  slot: number;
}

export interface Pokemon {
  abilities: Ability[];
  base_experience: number;
  forms: Species[];
  game_indices: GameIndex[];
  height: number;
  held_items: HeldItem[];
  id: number;
  is_default: boolean;
  location_area_encounters: string;
  moves: Move[];
  name: string;
  order: number;
  past_types: unknown[];
  species: Species;
  sprites: Sprites;
  stats: Stat[];
  types: Type[];
  weight: number;
}

export interface IPagination {
  firstPage: number;
  lastPage: number;
  nextPages: number[];
  prevPages: number[];
  nextPage: number;
  prevPage: number;
  page: number;
}
