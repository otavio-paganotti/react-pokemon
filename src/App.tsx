import routes from '@/routes';
import { useRoutes } from 'react-router-dom';
import '@/assets/css/main.scss';

const App = () => {
  const content = useRoutes(routes);

  return content;
};

export default App;
