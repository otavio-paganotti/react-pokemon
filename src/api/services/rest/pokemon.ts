import { http } from "@/api";
import { Pokemon, PokemonItem, Response } from "@/types";

export const getPokemon = async (id: number): Promise<Pokemon> => {
  const { data } = await http.get<Pokemon>(`/pokemon/${id}`);
  return data;
};

export const getPokemons = async (): Promise<Response<PokemonItem>> => {
  const { data } = await http.get<Response<PokemonItem>>("/pokemon");
  return data;
}