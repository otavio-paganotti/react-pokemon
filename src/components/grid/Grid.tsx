import Paginator from './Paginator';
import List from './List';
import { useLoader } from '@/hooks';
import { useEffect, useState } from 'react';
import { PokemonItem, Response } from '@/types';

const GenericGrid = <T extends PokemonItem>({
  module,
}: {
  module: (page?: number) => Promise<Response<T>>;
}) => {
  const [initial] = useState<Response<T>>({
    count: 0,
    next: null,
    previous: null,
    results: [],
    pagination: {
      firstPage: 1,
      lastPage: 1,
      nextPage: 1,
      prevPage: 1,
      nextPages: [1],
      prevPages: [1],
      page: 1,
    },
  });

  const [loader, data, isLoading] = useLoader<Response<T>>(initial);

  useEffect(() => {
    loader(module);

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const page = (page: number) => {
    if (isLoading) return;

    loader(() => module(page));
  };

  const handleSearch = () => {
    loader(() =>
      Promise.resolve({
        count: 0,
        next: null,
        previous: null,
        results: [],
        pagination: {
          firstPage: 0,
          lastPage: 0,
          nextPage: 0,
          prevPage: 0,
          nextPages: [0],
          prevPages: [0],
          page: 0,
        },
      })
    );
  };

  const handleClearSearch = () => {
    loader(module);
  };

  return (
    <div>
      {isLoading && <div>Loading...</div>}
      <div className="p-4 h-screen-minus-56 grid grid-cols-1 content-between gap-4">
        {!isLoading && (
          <>
            <div className="flex gap-4 justify-between">
              <input
                type="search"
                className="w-full px-4 py-2"
                placeholder="Busque o PokÃ©mon..."
              />
              <button
                className="px-4 py-2 text-white"
                onClick={handleSearch}
                style={{ backgroundColor: 'blue' }}
              >
                Buscar
              </button>
              <button
                className="px-4 py-2"
                onClick={handleClearSearch}
                style={{ backgroundColor: 'gray', color: 'black' }}
              >
                Limpar
              </button>
            </div>
            <List items={data?.results} />
          </>
        )}
        <Paginator pagination={data?.pagination} page={page} />
      </div>
    </div>
  );
};

export default GenericGrid;
