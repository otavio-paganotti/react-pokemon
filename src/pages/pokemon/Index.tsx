import { Pokemon as PokemonApi } from '@/api';
import Card from '@/components/ui/Card';
import { useLoader } from '@/hooks';
import { Pokemon } from '@/types';
import { useEffect } from 'react';
import { useParams } from 'react-router-dom';

export default function PokemonPage() {
  const params = useParams();

  const [loader, data, isLoading] = useLoader<Pokemon>({} as Pokemon);

  useEffect(() => {
    loader(() => PokemonApi.getPokemon(Number(params.id)));

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [params.id]);

  const getXAttack = () =>
    data.stats?.find((stat) => stat.stat.name === 'special-attack')?.base_stat;
  const getXDefense = () =>
    data.stats?.find((stat) => stat.stat.name === 'special-defense')?.base_stat;
  const getSpeed = () =>
    data.stats?.find((stat) => stat.stat.name === 'speed')?.base_stat;

  const getUrl = (id: number) =>
    `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/${id}.png`;

  return (
    <>
      <div className="p-4 max-w-3xl mx-auto grid grid-cols-1 gap-4">
        <h1 className="text-4xl text-center">Pokemon</h1>
        <p className="text-center">This is the Pokemon page</p>
        {!isLoading && (
          <>
            <Card
              item={{
                id: data.id,
                url: getUrl(data.id),
                name: data.name,
                content: data,
              }}
              noLink={true}
            >
              <div className="px-2 pb-2">
                <p>Ataque especial: {getXAttack()}</p>
                <p>Defesa especial: {getXDefense()}</p>
                <p>Velocidade: {getSpeed()}</p>
              </div>
            </Card>
          </>
        )}
      </div>
    </>
  );
}
