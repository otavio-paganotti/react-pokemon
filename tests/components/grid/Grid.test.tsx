import React from 'react';
import renderer, { act } from 'react-test-renderer';
import Grid from '../../../src/components/grid/Grid';
import { Pokemon as PokemonApi } from '../../../src/api';
import { expect, describe, test, vi, beforeEach } from 'vitest';
import { pokemon } from '../../mocks/pokemon';

vi.mock('../../../src/api');
vi.mock('react-router-dom');

const toJson = (component: renderer.ReactTestRenderer) => {
  const result = component.toJSON();
  expect(result).toBeDefined();
  expect(result).not.toBeInstanceOf(Array);
  return result as renderer.ReactTestRendererJSON;
};

const mockedResolvedResponse = {
  count: 0,
  next: null,
  previous: null,
  results: [pokemon],
  pagination: {
    firstPage: 1,
    lastPage: 1,
    nextPage: 1,
    prevPage: 1,
    nextPages: [],
    prevPages: [],
    page: 1,
  },
};

const onMounted = async () => {
  vi.mocked(PokemonApi.getPokemons).mockResolvedValueOnce(
    mockedResolvedResponse
  );

  let component;

  await act(async () => {
    component = renderer.create(<Grid module={PokemonApi.getPokemons} />);
  });

  return toJson(component);
};

describe('Grid.tsx', () => {
  let tree;

  beforeEach(async () => {
    tree = await onMounted();

    vi.mocked(PokemonApi.getPokemons).mockClear();
  });

  test('should render correctly the Grid component and take a snapshot', () => {
    expect(tree).toMatchSnapshot();
  });

  test('should Grid component have 1 children', () => {
    expect(tree.children?.length).toBe(1);
  });

  test('should Grid component have a div with class "p-4 h-screen-minus-56 grid grid-cols-1 content-between gap-4"', () => {
    expect(tree.children[0].props.className).toBe(
      'p-4 h-screen-minus-56 grid grid-cols-1 content-between gap-4'
    );
  });

  test('should list section have a list and a pagination component', () => {
    expect(tree.children[0].children?.length).toBe(2);
    expect(tree.children[0].children[0].type).toBe('div');
    expect(tree.children[0].children[1].type).toBe('div');
  });
});
