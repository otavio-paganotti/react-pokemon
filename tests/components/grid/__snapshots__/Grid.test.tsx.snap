// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Grid.tsx > should render correctly the Grid component and take a snapshot 1`] = `
<div>
  <div
    className="p-4 h-screen-minus-56 grid grid-cols-1 content-between gap-4"
  >
    <div
      className="grid grid-cols-1 md:grid-cols-4 gap-4"
    >
      <div
        className="bg-white"
      >
        <div
          className="flex justify-center"
        >
          <img
            src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png"
          />
        </div>
        <div
          className="p-4 flex flex-col gap-4"
        >
          <div
            className="bg-primary"
          >
            <div
              className="p-2 flex justify-between"
            >
              <h2
                className="font-bold capitalize"
              >
                bulbasaur
              </h2>
              <span>
                HP: 
                45
              </span>
            </div>
            <div
              className="px-2 pt-2 pb-2"
            >
              <p>
                Ataque: 
              </p>
              <p>
                Defesa: 
              </p>
            </div>
          </div>
          <div
            className="flex justify-center"
          />
        </div>
      </div>
    </div>
    <div
      className="flex justify-center mb-4"
    >
      <div
        className="mb-4"
      >
        <nav
          aria-label="Pagination"
          className="flex md:gap-4"
        >
          <div
            className="page-item"
            onClick={[Function]}
          >
            &lt;&lt;
          </div>
          <div
            className="page-item"
            onClick={[Function]}
          >
            &lt;
          </div>
          <div
            className="page-item current"
            onClick={[Function]}
          >
            1
          </div>
          <div
            className="page-item"
            onClick={[Function]}
          >
            &gt;
          </div>
          <div
            className="page-item"
            onClick={[Function]}
          >
            &gt;&gt;
          </div>
        </nav>
      </div>
    </div>
  </div>
</div>
`;
